### PR TITLE
feat(idempotency): add idempotency decorator

### LINKED ISSUE #1700: Feature request: Add back Idempotent decorators
### Use case

I want to annotate class methods with a decorator.
Example:
```ts
export class Foo {
  @MakeIdempotent(config)
  public submitPayment(): void {
    // Do something
  }
}
``` 

Since the package now is based on typescript 5.x (#1375) I'd assume this should not be a problem.

### Solution/User Experience

Add the decorators back in which have been removed with #1553

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda-typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

### PR DESCRIPTION
**Pull request by am29d**
## Description of your changes

This PR adds `@idempotent` decorator to the utility. We initially implemented it, but removed later due to uncertainty about decorator interface with TypeScript 5.x upgrade. Now we can restore and simplify the old implementation, by internally calling `makeIdempotent`, making it syntactic sugar for this functionality. This also means that most of e2e tests are very similar to the `makeIdempotent` tests, but we still need to make sure that it works when wrapping around decorator.

### Related issues, RFCs

**Issue number:** closes #1700 

## Checklist

- [x] [My changes meet the tenets criteria](https://docs.powertools.aws.dev/lambda-typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] My changes generate *no new warnings*
- [x] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: docs/snippets/idempotency/idempotentDecoratorBase.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import type { Context } from 'aws-lambda';
+import { LambdaInterface } from '@aws-lambda-powertools/commons';
```
Comment (dreamorosi) [inline docs/snippets/idempotency/idempotentDecoratorBase.ts:2]:
```suggestion
import type { LambdaInterface } from '@aws-lambda-powertools/commons';
```

File: docs/snippets/idempotency/idempotentDecoratorBase.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import {
+  IdempotencyConfig,
+  idempotent,
+} from '@aws-lambda-powertools/idempotency';
+import { DynamoDBPersistenceLayer } from '@aws-lambda-powertools/idempotency/dynamodb';
+import { Request, Response } from './types';
```
Comment (dreamorosi) [inline docs/snippets/idempotency/idempotentDecoratorBase.ts:8]:
```suggestion
import type { Request, Response } from './types';
```

File: docs/utilities/idempotency.md [hunk 3]
```diff
@@ -160,6 +160,22 @@ When using `makeIdempotent` on arbitrary functions, you can tell us which argume
...
+
+=== "index.ts"
+
+    ```typescript hl_lines="17"
+    --8<-- "docs/snippets/idempotency/idempotentDecoratorBase.ts"
+    ```
+
+=== "types.ts"
+
+    ```typescript
```
Comment (dreamorosi) [inline docs/utilities/idempotency.md:175]: Is this clipped intentionally? I think it's missing the link to the file & closing symbol.

File: docs/utilities/idempotency.md [hunk 4]
```diff
@@ -160,6 +160,22 @@ When using `makeIdempotent` on arbitrary functions, you can tell us which argume
...
+
+You can use the decorator on your Lambda handler or on any function that returns a response to make it idempotent. This is useful when you want to make a specific logic idempotent, for example when your Lambda handler performs multiple side effects and you only want to make a specific one idempotent.
```
Comment (dreamorosi) [inline docs/utilities/idempotency.md:177]: Suggest to change `on any function` -> `on any method`

**Comment by am29d** [comment]
✅ [e2e tests](https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/6340404317)

File: packages/idempotency/README.md [hunk 5]
```diff
@@ -24,7 +25,7 @@ You can use the package in both TypeScript and JavaScript code bases.
 ## Intro
 
 This package provides a utility to implement idempotency in your Lambda functions. 
-You can either use it to wrap a function, or as Middy middleware to make your AWS Lambda handler idempotent.
+You can either use it to wrap a function, decorate a function, or as Middy middleware to make your AWS Lambda handler idempotent.
```
Comment (dreamorosi) [inline packages/idempotency/README.md:28]: Same as above, I'd call this `method` rather than `function`.

File: packages/idempotency/README.md [hunk 6]
```diff
@@ -163,6 +164,69 @@ export const handler = makeIdempotent(myHandler, {
...
+  tableName: 'idempotencyTableName',
+});
+
+class MyHandler extends LambdaInterface {
+  @idempotent({ persistenceStore: dynamoDBPersistenceLayer })
+  public async handler(
+          event: APIGatewayProxyEvent,
+          context: Context
+  ): Promise<void> {
+    // your code goes here here
```
Comment (dreamorosi) [inline packages/idempotency/README.md:187]: I'd suggest to return something here.

File: packages/idempotency/README.md [hunk 7]
```diff
@@ -163,6 +164,69 @@ export const handler = makeIdempotent(myHandler, {
 
 Check the [docs](https://docs.powertools.aws.dev/lambda/typescript/latest/utilities/idempotency/) for more examples.
 
+### Decorator
+
+You can make any function idempotent, and safe to retry, by decorating it using the `@idempotent` decorator.
+
+```ts
+import { idempotent } from '@aws-lambda-powertools/idempotency';
+import { LambdaInterface } from '@aws-lambda-powertools/commons';
```
Comment (dreamorosi) [inline packages/idempotency/README.md:173]:
```suggestion
import type { LambdaInterface } from '@aws-lambda-powertools/commons';
```

File: packages/idempotency/README.md [hunk 8]
```diff
@@ -163,6 +164,69 @@ export const handler = makeIdempotent(myHandler, {
...
+import type { Context, APIGatewayProxyEvent } from 'aws-lambda';
+
+const persistenceStore = new DynamoDBPersistenceLayer({
+  tableName: 'idempotencyTableName',
+});
+
+class MyHandler extends LambdaInterface {
+  @idempotent({ persistenceStore: dynamoDBPersistenceLayer })
+  public async handler(
+          event: APIGatewayProxyEvent,
```
Comment (dreamorosi) [inline packages/idempotency/README.md:184]: Minor: not 100% sure, but I think the indent is off here

File: packages/idempotency/README.md [hunk 9]
```diff
@@ -163,6 +164,69 @@ export const handler = makeIdempotent(myHandler, {
...
+  }
+}
+
+const handlerClass = new MyHandler();
+export const handler = handlerClass.handler.bind(handlerClass);
+```
+
+Using the same decorator, you can also make any other arbitrary function idempotent.
+
+```ts
```
Comment (dreamorosi) [inline packages/idempotency/README.md:197]: Same comments from the previous snippet above apply here.

File: packages/idempotency/src/idempotencyDecorator.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Similar to decoratoring a handler you can use the decorator on any other function.
+ * @example
+ * ```ts
+ * import {
+ *  DynamoDBPersistenceLayer,
+ *  idempotentFunction
+ *  } from '@aws-lambda-powertools/idempotency'
+ *
+ *  class MyClass {
```
Comment (dreamorosi) [inline packages/idempotency/src/idempotencyDecorator.ts:34]: I think both snippets are missing the `implements LambdaInterface` and might be missing some closing braces.

**Review by dreamorosi — APPROVED** [review]
Great work on this PR, Alex!

I've left a few comments on the docs only. The general feedback would be to: 1/ let's try to start using `import type { fooBar } from '...';` when importing types. This helps with telling TS what we are importing.

**Comment by dreamorosi** [comment]
Thanks for running the integration tests, they are passing 🎉 

Leaving the link here for the record: https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/6340404317

File: packages/idempotency/README.md (hunk 6 above)
Comment (am29d) [reply packages/idempotency/README.md:187]: Nah, let's keep it consistent with `makeIdempotent`

**Pull request merged**